File: sdk/src/main/java/eu/darken/porter/sdk/PorterServiceConnection.java
package eu.darken.porter.sdk;

import android.content.ComponentName;
import android.content.ServiceConnection;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.os.RemoteException;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.EnumMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

class PorterServiceConnection implements UserServiceCallback {

    private static final Handler MAIN_HANDLER = new Handler(Looper.getMainLooper());

    private final Set<ServiceConnection> connections = new HashSet<>();
    /** Guarded by this instance, which is also what a wire locks while it registers one. */
    private final Map<PorterBackend, IBinder> registeredBinders = new EnumMap<>(PorterBackend.class);
    private final ComponentName componentName;
    private IBinder binder;

    public PorterServiceConnection(Porter.UserServiceArgs args) {
        this.componentName = args.componentName;
    }

    private boolean dead = false;
    /** Counts the changes to {@link #connections}, so a queued death can tell it is stale. */
    private int generation = 0;

    /** @return whether {@code conn} was not registered already */
    public boolean addConnection(@Nullable ServiceConnection conn) {
        if (conn == null || !connections.add(conn)) return false;
        generation++;
        // A registration made after a death is a live binding again, and is owed a later one.
        dead = false;
        return true;
    }

    public void removeConnection(@Nullable ServiceConnection conn) {
        if (conn != null) {
            connections.remove(conn);
        }
    }

    public void clearConnections() {
        generation++;
        connections.clear();
    }

    @Nullable
    @Override
    public IBinder registeredBinder(@NonNull PorterBackend backend) {
        synchronized (this) {
            return registeredBinders.get(backend);
        }
    }

    @Override
    public void rememberRegisteredBinder(@NonNull PorterBackend backend, @NonNull IBinder binder) {
        synchronized (this) {
            registeredBinders.put(backend, binder);
        }
    }

    @Override
    public void connected(@NonNull IBinder binder) {
        MAIN_HANDLER.post(() -> {
                    for (ServiceConnection conn : connections) {
                        conn.onServiceConnected(componentName, binder);
                    }
                }
        );

        // Hold the binder, or linkToDeath will not work after reference to
        // the binder is dropped
        this.binder = binder;

        try {
            this.binder.linkToDeath(this::died, 0);
        } catch (RemoteException ignored) {
        }
    }

    @Override
    public void died() {
        binder = null;

        if (dead) return;
        dead = true;

        // A rebind can land before this runs. It is then a binding of its own, and the death
        // belongs to the one it replaced: telling it would disconnect a caller nothing happened to.
        int atDeath = generation;
        MAIN_HANDLER.post(() -> {
                    if (generation != atDeath) return;

                    for (ServiceConnection conn : connections) {
                        conn.onServiceDisconnected(componentName);
                    }

                    connections.clear();
                    PorterServiceConnections.remove(this);
                }
        );
    }
}

File: sdk/src/main/java/eu/darken/porter/sdk/PorterServiceConnection.kt
package eu.darken.porter.sdk

import android.content.ComponentName
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import android.os.RemoteException
import java.util.EnumMap

internal class PorterServiceConnection(
    private val registry: PorterServiceConnections,
    args: UserServiceArgs,
) : UserServiceCallback {

    /** Guarded by [PorterServiceConnections.lock]. */
    private val listeners = HashSet<UserServiceListener>()

    /** Guarded by this instance, which is also what a wire locks while it registers one. */
    private val registeredBinders = EnumMap<PorterBackend, IBinder>(PorterBackend::class.java)

    private val componentName: ComponentName = args.componentName

    private var binder: IBinder? = null

    /** Set once, by the death of the binder this was bound to. Guarded by the same lock. */
    private var terminal = false

    /** The set a queued death delivery will be handed, null once it ran or was called off. */
    private var pendingDelivery: List<UserServiceListener>? = null

    /** @return whether [listener] was inserted, so a refused call can remove what it added */
    fun addListener(listener: UserServiceListener?): Boolean = synchronized(registry.lock) {
        listener != null && listeners.add(listener)
    }

    fun removeListener(listener: UserServiceListener?) {
        synchronized(registry.lock) {
            if (listener != null) listeners.remove(listener)
        }
    }

    /** Whether anyone is still registered here. The caller holds [PorterServiceConnections.lock]. */
    fun hasListeners(): Boolean = listeners.isNotEmpty()

    /** Whether the binder this was bound to has died. The caller holds [PorterServiceConnections.lock]. */
    fun isTerminal(): Boolean = terminal

    fun clearListeners() {
        synchronized(registry.lock) {
            listeners.clear()
        }
    }

    override fun registeredBinder(backend: PorterBackend): IBinder? = synchronized(this) {
        registeredBinders[backend]
    }

    override fun rememberRegisteredBinder(backend: PorterBackend, binder: IBinder) {
        synchronized(this) {
            registeredBinders[backend] = binder
        }
    }

    override fun connected(binder: IBinder) {
        synchronized(registry.lock) {
            // Nothing is registered here any more and nothing can be, so holding the binder and
            // linking a recipient would only keep both alive for a delivery that cannot happen.
            if (terminal) return
        }

        MAIN_HANDLER.post {
            val snapshot = synchronized(registry.lock) { listeners.toList() }
            for (listener in snapshot) {
                listener.onConnected(componentName, binder)
            }
        }

        // Hold the binder, or linkToDeath will not work after reference to
        // the binder is dropped
        this.binder = binder

        try {
            binder.linkToDeath({ died() }, 0)
        } catch (ignored: RemoteException) {
        }
    }

    /** Calls off a delivery that has not run yet, for a caller that no longer wants the binding. */
    fun cancelPendingDelivery() {
        synchronized(registry.lock) {
            pendingDelivery = null
        }
    }

    override fun died() {
        binder = null

        synchronized(registry.lock) {
            // One binder carries a recipient per "connected" push, so its death arrives repeatedly.
            if (terminal) return
            terminal = true
            pendingDelivery = listeners.toList()
            listeners.clear()
            registry.retire(this)

            // Queued while the eviction is still private to this thread: a rebind has to take this
            // same lock to register, so its "connected" cannot reach the queue ahead of this
            // disconnect. Posting is not executing, the body runs later and outside the lock.
            MAIN_HANDLER.post {
                val snapshot = synchronized(registry.lock) {
                    val delivery = pendingDelivery
                    pendingDelivery = null
                    registry.deliveryDone(this)
                    delivery
                } ?: return@post

                for (listener in snapshot) {
                    listener.onDisconnected(componentName)
                }
            }
        }
    }

    private companion object {
        val MAIN_HANDLER = Handler(Looper.getMainLooper())
    }
}
